lib/streamlit/commands/execution_control.py

```diff
@@ -87,3 +90,51 @@ def experimental_rerun() -> NoReturn:
     # be seen.
     _LOGGER.warning(msg)
     rerun()
+
+
+@gather_metrics("switch_page")
+def switch_page(page: str) -> NoReturn:  # type: ignore[misc]
+    """Switch the current programmatically page in a multi-page app.
+    When `st.switch_page()` is called with a page, the current page script is halted
+    and the requested page script will be queued to run from the top.
+    Parameters
+    ----------
+    page: str
+        The file path, relative to the main script, of the page to switch to.
+    """
+
+    ctx = get_script_run_ctx()
+
+    ctx_main_script = ""
+    if ctx:
+        ctx_main_script = ctx.main_script_path
+
+    main_script_path = os.path.join(os.getcwd(), ctx_main_script)
+    main_script_directory = os.path.dirname(main_script_path)
+
+    # Convenience for handling ./ notation and ensure leading / doesn't refer to root directory
+    page = os.path.normpath(page.strip("/"))
+
+    # Build full path
+    requested_page = os.path.join(main_script_directory, page)
+    all_app_pages = source_util.get_pages(ctx_main_script).values()
+
+    page_found = False
+    for page_data in all_app_pages:
+        full_path = page_data["script_path"]
+
+        if requested_page == full_path and ctx and ctx.script_requests:
+            page_found = True
+            ctx.script_requests.request_rerun(
+                RerunData(
+                    query_string="",
+                    page_script_hash=page_data["page_script_hash"],
+                )
+            )
+            # Force a yield point so the runner can do the rerun
+            st.empty()
+
+    if not page_found:
+        raise StreamlitAPIException(
+            f"Could not find page: '{page}'. Must be the file path relative to the main script, from the directory: {os.path.basename(main_script_directory)}."
+        )
```

Comment:
I think all the following is good and clean. I might want to offer an idea that you can take or leave.

```python

if not ctx or not ctx.script_requests:
  # Log Warning
  return

matched_pages = [page_data for page_data in all_app_pages if page_data["script_path"] === requested_page]

if len(matched_pages) == 0:
  raise StreamlitAPIException(
            f"Could not find page: '{page}'. Must be the file path relative to the main script, from the directory: {os.path.basename(main_script_directory)}."
        )

ctx.script_requests.request_rerun(
  RerunData(
    query_string="",
    page_script_hash=matched_pages[0]["page_script_hash"],
  )
)
# Force a yield point so the runner can do the rerun
st.empty()
```

Note I haven't tested this, so you can verify. I tried to match the variable names to what you have for simplicity, but since this is a Python list comprehension, you can probably use shorter var names.

```python
matched_pages = [p for p in all_app_pages if p["script_path"] === requested_page]
```